pkg/cmd/roachtest/tests/restart_rolling_and_rolling.go

```diff
@@ -0,0 +1,115 @@
+// included in the /LICENSE file.
+
+package tests
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/cockroachdb/cockroach/pkg/cmd/roachtest/cluster"
+	"github.com/cockroachdb/cockroach/pkg/cmd/roachtest/option"
+	"github.com/cockroachdb/cockroach/pkg/cmd/roachtest/registry"
+	"github.com/cockroachdb/cockroach/pkg/cmd/roachtest/roachtestutil"
+	"github.com/cockroachdb/cockroach/pkg/cmd/roachtest/spec"
+	"github.com/cockroachdb/cockroach/pkg/cmd/roachtest/test"
+	"github.com/cockroachdb/cockroach/pkg/roachprod/install"
+)
+
+func runRestartRollingAndRolling(ctx context.Context, t test.Test, c cluster.Cluster) {
+	t.Status("starting the cluster")
+	startOpts := option.DefaultStartOpts()
+	cs := install.MakeClusterSettings()
+	cs.ClusterSettings["server.consistency_check.interval"] = "1m"
+	cs.ClusterSettings["server.consistency_check.max_rate"] = "1GB"
+	c.Put(ctx, "./cockroach-new", "./cockroach-new", c.CRDBNodes())
+	c.Start(ctx, t.L(), startOpts, install.MakeClusterSettings(), c.CRDBNodes())
+
+	// Limit the disk throughput to 128 MiB/s, to more easily stress IO overload.
+	const diskBand = 128 << 20 // 128 MiB
+	t.Status(fmt.Sprintf("limiting disk bandwidth to %d bytes/s", diskBand))
+	staller := roachtestutil.MakeCgroupDiskStaller(t, c,
+		false /* readsToo */, false /* logsToo */)
+	staller.Unstall(ctx, c.CRDBNodes())
+	staller.Setup(ctx)
+
+	t.Status("initializing kv workload")
+	// Init KV workload with a bunch of pre-split ranges and pre-inserted rows.
+	// The block sizes are picked the same as for the workload below.
+	c.Run(ctx, option.WithNodes(c.WorkloadNode()), fmt.Sprintf(
+		"./cockroach workload init kv --drop --splits=30000 --insert-count=10000 "+
+			"--min-block-bytes=128 --max-block-bytes=256 {pgurl%s}",
+		c.Node(1)))
+
+	staller.Slow(ctx, c.CRDBNodes(), diskBand)
+	defer staller.Unstall(ctx, c.CRDBNodes())
+
+	m := c.NewMonitor(ctx, c.CRDBNodes())
+	cancelWorkload := m.GoWithCancel(func(ctx context.Context) error {
+		t.Status("running kv workload")
+		const duration = 24 * time.Hour // basically infinite, we're relying on ctx getting canceled
+		// TODO: Tune the workload are to keep the cluster busy 60% CPU, and IO
+		// overload metric approaching 10-20%.
+		cmd := roachtestutil.NewCommand("./cockroach workload run kv "+
+			"--histograms=perf/stats.json --concurrency=500 "+
+			"--max-rate=5000 --read-percent=5 "+
+			"--min-block-bytes=128 --max-block-bytes=256 "+
+			"--txn-qos='regular' --tolerate-errors "+
+			"--duration=%v {pgurl%s}", duration, c.CRDBNodes())
+		err := c.RunE(ctx, option.WithNodes(c.WorkloadNode()), cmd.String())
+		if ctx.Err() != nil {
+			return nil // happy case
+		}
+		return err // workload returned actual error
+	})
+
+	restart := func(transitionToNew bool) {
+		for _, n := range c.CRDBNodes() {
+			m.ExpectDeath()
+			t.Status(fmt.Sprintf("Stopping node %d.", n))
+			c.Stop(ctx, t.L(), option.DefaultStopOpts(), c.Node(n))
```

Comment:
As a heads up, I this does a SIGKILL.  If we also want to test a mix of graceful shutdowns, we could do something like we do in some other tests:

```
	shouldUseSigKill := cfg.rng.Float64() > 0.5
	if shouldUseSigKill {
		t.L().Printf(`stopping node (using SIGKILL) %s`, target)
		if err := c.StopE(ctx, t.L(), option.DefaultStopOpts(), target); err != nil {
			return errors.Wrapf(err, "could not stop node %s", target)
		}
	} else {
		t.L().Printf(`stopping node gracefully %s`, target)
		if err := c.StopE(
			ctx, t.L(), option.NewStopOpts(option.Graceful(shutdownGracePeriod)), c.Node(cfg.shutdownNode),
		); err != nil {
			return errors.Wrapf(err, "could not stop node %s", target)
		}
	}
```